src/service.js

```diff
@@ -443,6 +498,35 @@ function getServiceInternal(holder, context, id, opt_factory,
   return s.obj;
 }
 
+/**
+ * @param {!Object} holder Object holding the service instance.
+ * @param {!Window|!./service/ampdoc-impl.AmpDoc} context Win or AmpDoc.
+ * @param {string} id of the service.
+ * @param {?function(new:Object, ?)=} opt_ctor
+ *     Constructor function to new the service. Called with context.
+ * @param {?function(?)=} opt_factory
+ *     Factory function to create the new service. Called with context.
+ */
+function registerServiceInternal(holder, context, id, opt_factory, opt_ctor) {
+  dev().assert(!!opt_factory != !!opt_ctor,
+    `Provide a constructor or a factory, but not both for service ${id}`);
+  const services = getServices(holder);
+  let s = services[id];
+  if (s && !s.obj && !s.ctor && !s.factory) {
```

Comment:
Long chains of logical ops are not very readable. I'd rewrite this like:
```
if (!s) {
  s = services[id] = ...
}
s.ctor = s.ctor || opt_ctor;
s.factory = s.factory || opt_factory;
```
The above code won't overwrite existing ctor/factory props. I recall you mention checking for that case?